### ISSUE TITLE
includeIf not working with WSL/Network paths

### ISSUE THREAD

**Issue by Stanzilla** [git-for-windows/git#2563]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.26.0.windows.1
cpu: x86_64
built from commit: 9c98e1ccdfd839e4eaae1c2747d0088ef89d446b
sizeof-long: 4
sizeof-size_t: 8
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19041.172]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VisualStudioCode
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: LFOnly
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I'm using this setup

```
; personal config
[includeIf "gitdir/i:T:/home/stan/projects/personal/"]
    path = .gitconfig-personal

[includeIf "gitdir/i:C:/Users/Stan/projects/personal/"]
    path = .gitconfig-personal

; work config
[includeIf "gitdir/i:T:/home/stan/projects/work/"]
    path = .gitconfig-work

[includeIf "gitdir/i:C:/Users/Stan/projects/work/"]
    path = .gitconfig-work
```

and the config files get correctly included for projects on the C drive, but not for projects on the T drive. T is a WSL mount. Is there a known issue about WSL compat or am I just doing something wrong?

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

PowerShell / Cmd

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git config user.email
```
 - What did you expect to occur after running these commands?

the correct email

 - What actually happened instead?

no email

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

N/A

**Comment by dscho** [comment]
> T is a WSL mount.

What is the output of `git rev-parse --absolute-git-dir` in `T:/home/stan/project/work/`?

A recent Git for Windows version started to resolve symbolic links/mount points/whatever when trying to canonicalize absolute directories, and that most likely changed the behavior of the `includeIf` for you.

**Comment by Stanzilla** [comment]
That gives `//wsl$/Ubuntu/home/stan/.git`

**Comment by dscho** [comment]
Those two paths look different to me... one ends in `/.git`, the other doesn't...

**Comment by Stanzilla** [comment]
According to the docs (https://git-scm.com/docs/git-config#_conditional_includes) this should work though? I want every folder inside of /work/ to use the work gitconfig etc.

**Comment by dscho** [comment]
Maybe you can build Git for Windows, start it in `gdb` and set a breakpoint on the relevant code?

**Comment by dscho** [comment]
https://github.com/git-for-windows/git/blob/9c98e1ccdfd839e4eaae1c2747d0088ef89d446b/config.c#L244 should be the best location for said breakpoint.

**Comment by Stanzilla** [comment]
that's sadly a bit out of my capabilities

**Comment by Stanzilla** [comment]
This is sadly still an issue, is there any other way I can help get this fixed without providing a PR?

**Comment by dscho** [comment]
> ```ini
> ; work config
> [includeIf "gitdir/i://wsl$/Ubuntu/home/stan/projects/work/"]
>     path = .gitconfig-work
> ```

I should have spotted this right away, sorry. There are two problems with this:

- this is not the path to the gitdir, it is the path to the top-level directory of the main worktree. If you append `/.git`, it should work
- the path you provide is relative, which in the case of `includeIf.<cond>.path` means: relative to the gitdir. Since you most likely put this file into the worktree, you need to prefix it with `../`. Note that Git will not complain about a missing file, giving you no indication that you used an incorrect path.

In short, I think that this should work for you:

```ini
; work config
[includeIf "gitdir/i://wsl$/Ubuntu/home/stan/projects/work/.git"]
    path = ../.gitconfig-work
```

**Comment by dscho** [comment]
> In short, I think that this should work for you:
> 
> ```ini
> ; work config
> [includeIf "gitdir/i://wsl$/Ubuntu/home/stan/projects/work/.git"]
>     path = ../.gitconfig-work
> ```

Whoops, I forgot one crucial thing. For technical reasons, Git for Windows assumes paths starting with a single slash to be relative to the MSYS2 root (i.e. the directory into which Git for Windows was installed, usually `C:\Program Files\Git`). If really a single slash was meant, it has to be doubled, as per MSYS2's convention. Which means that your `//wsl$` is mistaken for meaning `/wsl$`. You need to use _three_ slashes:

```ini
; work config
[includeIf "gitdir/i:///wsl$/Ubuntu/home/stan/projects/work/.git"]
    path = ../.gitconfig-work
```

> * this is not the path to the gitdir, it is the path to the top-level directory of the main worktree

Actually, that is an assumption on my part that might not even be true, if you have multiple worktrees inside that directory.

Note, however, that the `gitdir/i:<pattern>` condition uses a wildcard pattern, read: you could say `"gitdir/i://wsl$/Ubuntu/home/stan/projects/work/**"`.

Also note that the `path` value, if relative, is actually relative to the directory in which the config file lives that contains that `includeIf` setting (I said previously that it is relative to the gitdir, which is incorrect). That is, if you configure this in `~/.gitconfig` (which I assume you do, otherwise you would already know precisely what your `gitdir` is and would not have to have a conditional include), the value `.gitconfig-work` would mean that the file `~/.gitconfig-work` would be read.

**Comment by Stanzilla** [comment]
Okay thank you for the explanation, my head hurts a bit but that is fine.

So my setup now is:

`.gitconfig` in `C:\Users\Stan`

This contains:

```config
[includeIf "gitdir/i:///wsl$/Ubuntu/home/stan/projects/personal/**"]
    path = .gitconfig-personal

[includeIf "gitdir/i:///wsl$/Ubuntu/home/stan/projects/work/**"]
    path = .gitconfig-work
```

Both `.gitconfig-personal` and `.gitconfig-work` are also in `C:\Users\Stan`.

So, now every project in `home/stan/projects/personal/` should use the personal config and the ones in work, the work one, correct?

**Comment by dscho** [comment]
> So, now every project in `home/stan/projects/personal/` should use the personal config and the ones in work, the work one, correct?

Correct.

> At least on the command line, this seems to work!

Great!

> The extra `/` is definitely what tripped me up here, so thank you so much for clearing that up.

Yes, I can see how that is confusing, especially since you think you're working in WSL, but then, you still use Git _for Windows_ ;-)

There is probably a lot of room for improvement when it comes to offering debugging capabilities in the `includeIf` machinery. Please do feel free to open a ticket at https://github.com/gitgitgadget/git/issues/new about this (not here, because it is not a Windows-specific issue).

As to this here ticket: I think it is resolved, so I will close it.
